copier/settings.py

```diff
@@ -0,0 +1,61 @@
+"""User setttings models and helper functions."""
+
+from __future__ import annotations
+
+import os
+import warnings
+from os.path import expanduser
+from pathlib import Path
+from typing import Any
+
+import yaml
+from platformdirs import user_config_path
+from pydantic import BaseModel, Field
+
+from .errors import MissingSettingsWarning
+
+ENV_VAR = "COPIER_SETTINGS_PATH"
+
+
+class Settings(BaseModel):
+    """User settings model."""
+
+    defaults: dict[str, Any] = Field(
+        default_factory=dict, description="Default values for questions"
+    )
+    trust: set[str] = Field(
+        default_factory=set, description="List of trusted repositories or prefixes"
+    )
+
+    @classmethod
+    def from_file(cls, settings_path: Path | None = None) -> Settings:
+        """Load settings from a file."""
+        env_path = os.getenv(ENV_VAR)
+        if settings_path is None:
+            if env_path:
+                settings_path = Path(env_path)
+            else:
+                settings_path = user_config_path("copier") / "settings.yml"
+        if settings_path.is_file():
+            data = yaml.safe_load(settings_path.read_text())
+            return cls.model_validate(data)
+        elif env_path:
+            warnings.warn(
+                f"Settings file not found at {env_path}", MissingSettingsWarning
+            )
+        return cls()
+
+    def is_trusted(self, repository: str) -> bool:
+        """Check if a repository is trusted."""
+        return any(
+            repository.startswith(self.normalize(trusted))
+            if trusted.endswith("/")
+            else repository == self.normalize(trusted)
+            for trusted in self.trust
+        )
+
+    def normalize(self, url: str) -> str:
+        """Normalize an URL using user settings."""
+        if url.startswith("~"):  # Only expand on str to avoid messing with URLs
+            url = expanduser(url)
+        return url
```

Comment:
I wonder whether we should use [`copier.vcs.get_repo()`](https://github.com/copier-org/copier/blob/557c0d614497e9b1b3382d4fd4be0d773d612bff/copier/vcs.py#L94-L127) to normalize trusted repository URLs to be more robust against non-canonical URLs. From the tests below, a (normalized) repository `https://github.com/user/repo.git` wouldn't be considered trusted despite the trust list entry `https://github.com/user/repo`. Won't this lead to confusion?

I realize that `copier.vcs.get_repo()` currently doesn't handle URLs with a trailing slash properly:

```pycon
>>> get_repo("https://github.com/user/repo/")
'https://github.com/user/repo/.git'
```

But this could be fixed. WDYT?

Reply:
Yes, I noticed that when I started #1941. That's one of the reasons I separated the shortcuts into their own PR.
If this is OK for you, I would prefer handling `get_repo` and URL normalization in #1941 as it is its main point. It would allow having the full picture and craft a solution that handles more cases.

I think we have the opportunity to properly review/complete/fix normalization rules in #1941

Reply:
Okay, let's address URL normalization in #1941.